src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs

```diff
@@ -81,21 +83,77 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
                 callingConvention,
                 refKind,
                 returnType,
+                refReadonlyModifiers,
                 syntax,
                 typeBinder,
                 diagnostics,
                 suppressUseSiteDiagnostics);
         }
 
+        public static FunctionPointerMethodSymbol CreateFromMetadata(CallingConvention callingConvention, ImmutableArray<ParamInfo<TypeSymbol>> retAndParamTypes)
+        {
+            Debug.Assert(retAndParamTypes.Length > 0);
+
+            ParamInfo<TypeSymbol> retInfo = retAndParamTypes[0];
+            var returnType = TypeWithAnnotations.Create(retInfo.Type, customModifiers: CSharpCustomModifier.Convert(retInfo.CustomModifiers));
+            var refCustomMods = CSharpCustomModifier.Convert(retInfo.RefCustomModifiers);
+
+            RefKind refKind = retInfo.IsByRef switch
+            {
+                false => RefKind.None,
+                true when CustomModifierUtils.HasInAttributeModifier(refCustomMods) => RefKind.RefReadOnly,
+                true => RefKind.Ref,
+            };
+
+            return new FunctionPointerMethodSymbol(
+                callingConvention,
+                refKind,
+                refCustomMods,
+                returnType,
+                retAndParamTypes.AsSpan()[1..]);
+        }
+
+        private static ImmutableArray<FunctionPointerParameterSymbol> MakeParametersFromMetadata(ReadOnlySpan<ParamInfo<TypeSymbol>> parameterTypes, FunctionPointerMethodSymbol parent)
+        {
+            if (parameterTypes.Length > 0)
+            {
+                var paramsBuilder = ArrayBuilder<FunctionPointerParameterSymbol>.GetInstance(parameterTypes.Length);
+
+                for (int i = 0; i < parameterTypes.Length; i++)
+                {
+                    ParamInfo<TypeSymbol> param = parameterTypes[i];
+                    var paramRefCustomMods = CSharpCustomModifier.Convert(param.RefCustomModifiers);
+                    var paramType = TypeWithAnnotations.Create(param.Type, customModifiers: CSharpCustomModifier.Convert(param.CustomModifiers));
+                    // PROTOTYPE(func-ptr): Need to encode out params as a custom modifier of some kind
+                    var paramRefKind = param.IsByRef switch
+                    {
+                        false => RefKind.None,
+                        true when CustomModifierUtils.HasInAttributeModifier(paramRefCustomMods) => RefKind.In,
+                        true => RefKind.Ref,
+                    };
```

Comment:
Consider extracting a helper method for use here and in `CreateFromMetadata()` above.